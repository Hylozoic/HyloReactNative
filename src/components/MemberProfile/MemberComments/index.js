import component from './MemberPosts'
import connector from './MemberPosts.connector'
export default connector(component)
