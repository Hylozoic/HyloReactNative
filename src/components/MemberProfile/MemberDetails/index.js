import { withNavigationFocus } from 'react-navigation-is-focused-hoc'

import component from './MemberDetails'
import connector from './MemberDetails.connector'
export default withNavigationFocus(connector(component))
