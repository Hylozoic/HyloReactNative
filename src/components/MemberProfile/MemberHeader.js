import React from 'react'
import { View, TouchableOpacity, TextInput, Text } from 'react-native'
import Icon from '../Icon'
import EntypoIcon from 'react-native-vector-icons/Entypo'
import PopupMenuButton from '../PopupMenuButton'
import { filter, isEmpty } from 'lodash/fp'
import styles from './MemberHeader.styles'

export default function MemberHeader ({
  person,
  flagMember,
  onPressMessages,
  isMe,
  editProfile,
  editable,
  updateSetting = () => {},
  saveChanges,
  errors = {}
}) {
  if (!person) return null

  const { name, location, tagline } = person
  return <View style={styles.header}>
    <View style={styles.nameRow}>
      <Control
        style={styles.name}
        value={name}
        editable={editable}
        onChangeText={updateSetting('name')}
        error={errors.name} />
      <View style={styles.icons}>
        <TouchableOpacity onPress={onPressMessages}>
          <Icon name='Messages' style={styles.icon} />
        </TouchableOpacity>
        <MemberMenu {... {flagMember, isMe, editProfile, saveChanges, editable}} />
      </View>
    </View>
    <Control
      style={styles.location}
      value={location}
      editable={editable}
      onChangeText={updateSetting('location')} />
    <Control
      style={styles.tagline}
      value={tagline}
      editable={editable}
      onChangeText={updateSetting('tagline')} />
  </View>
}

export class Control extends React.Component {
  focus = () => {
    this.input && this.input.focus()
  }

  render () {
    const { value, onChangeText, editable, style, onBlur, multiline, hideEditIcon, error } = this.props
    return <View style={[styles.control, editable && styles.editableControl]}>
      <View style={styles.controlInputRow}>
        <TextInput
          ref={i => { this.input = i }}
          style={[styles.controlInput, style]}
          value={value}
          onChangeText={onChangeText}
          editable={editable}
          onBlur={onBlur}
          multiline={multiline} />
        {editable && !hideEditIcon && <TouchableOpacity onPress={this.focus} style={styles.editIconWrapper}>
          <EntypoIcon name='edit' style={styles.editIcon} />
        </TouchableOpacity>}
      </View>
      {!!error && <View style={styles.controlError}><Text style={styles.controlErrorText}>{error}</Text></View>}
    </View>
  }
}

export function MemberMenu ({flagMember, isMe, editProfile, saveChanges, editable}) {
  // If the function is defined, than it's a valid action
  const actions = filter(x => x[1], [
    ['Edit', isMe && !editable && editProfile],
    ['Save Changes', isMe && editable && saveChanges],
    ['Flag This Member', !isMe && flagMember]
  ])

  if (isEmpty(actions)) return null

  const onSelect = index => actions[index][1]()

  const destructiveButtonIndex = actions[0][0] === 'Flag This Member' ? 0 : -1

  return <PopupMenuButton actions={actions.map(x => x[0])} onSelect={onSelect}
    destructiveButtonIndex={destructiveButtonIndex}>
    <Icon name='More' style={styles.lastIcon} />
  </PopupMenuButton>
}
