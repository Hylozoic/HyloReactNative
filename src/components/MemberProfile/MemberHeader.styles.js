import {
  capeCod, rhino80, rhino60, rhino50, amaranth
} from '../../style/colors'

export default {
  header: {
    marginBottom: 15
  },
  nameRow: {
    flexDirection: 'row',
    marginBottom: 5
  },
  name: {
    fontSize: 24,
    color: capeCod,
    fontFamily: 'Circular-Bold'
  },
  icons: {
    flexDirection: 'row',
    marginLeft: 'auto',
    paddingLeft: 10
  },
  icon: {
    fontSize: 30,
    color: rhino60,
    marginRight: 10
  },
  lastIcon: {
    fontSize: 30,
    color: rhino60
  },
  location: {
    fontSize: 14,
    color: rhino50,
    fontFamily: 'Circular-Book',
    marginBottom: 2
  },
  tagline: {
    fontSize: 16,
    color: rhino80,
    fontFamily: 'Circular-Book'
  },
  control: {
    flex: 1
  },
  editableControl: {
    marginBottom: 5
  },
  controlInputRow: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  controlInput: {
    textAlignVertical: 'top',
    flex: 1
  },
  editIconWrapper: {
    marginRight: 10
  },
  editIcon: {
    color: rhino60,
    fontSize: 16
  },
  controlError: {
    paddingTop: 5,
    paddingBottom: 3
  },
  controlErrorText: {
    fontSize: 13,
    fontFamily: 'Circular-Book',
    color: amaranth
  }
}
