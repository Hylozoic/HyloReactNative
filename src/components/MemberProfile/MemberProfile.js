import React from 'react'
import { Text, View, Image, TouchableOpacity } from 'react-native'
import Icon from '../Icon'
import Loading from '../Loading'
import styles from './MemberProfile.styles'
import MemberFeed from './MemberFeed'

export default class MemberProfile extends React.Component {
  static navigationOptions = () => ({
    headerTitle: 'Member'
  })

  componentDidMount () {
    this.props.fetchPerson()
  }

  componentDidUpdate (prevProps) {
    if (prevProps.id !== this.props.id) {
      this.props.fetchPerson()
    }
  }

  render () {
    const { person, id, goToDetails } = this.props

    if (!person) return <Loading />

    const header = <View>
      <MemberBanner person={person} />
      <View style={styles.marginContainer}>
        <MemberHeader person={person} />
        <ReadMoreButton goToDetails={goToDetails} />
      </View>
    </View>

    return <MemberFeed id={id} header={header} />
  }
}

export function MemberBanner ({ person }) {
  const { bannerUrl, avatarUrl } = person
  return <View>
    <Image source={{uri: bannerUrl}} style={styles.bannerImage} />
    <View style={styles.avatarWrapper}>
      <Image source={{uri: avatarUrl}} style={styles.avatarImage} />
    </View>
  </View>
}

export function MemberHeader ({ person }) {
  if (!person) return null
  const { name, location, tagline } = person
  return <View style={styles.header}>
    <View style={styles.nameRow}>
      <Text style={styles.name}>{name}</Text>
      <View style={styles.icons}>
        <Icon name='Messages' style={styles.icon} />
        <Icon name='More' style={styles.lastIcon} />
      </View>
    </View>
    <Text style={styles.location}>{location}</Text>
    <Text style={styles.tagline}>{tagline}</Text>
  </View>
}

export function ReadMoreButton ({ goToDetails }) {
  return <View style={styles.buttonContainer}>
    <TouchableOpacity onPress={goToDetails} style={styles.buttonWrapper}>
      <View style={styles.button}>
        <Text style={styles.buttonText}>Read More</Text>
      </View>
    </TouchableOpacity>
  </View>
}
