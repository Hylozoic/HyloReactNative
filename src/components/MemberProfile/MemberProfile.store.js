import { createSelector as ormCreateSelector } from 'redux-orm'
import orm from 'store/models'

export const MODULE_NAME = 'MemberProfile'
export const CLEAR_FETCH_PERSON_PENDING = `${MODULE_NAME}/CLEAR_FETCH_PERSON_PENDING`
export const FETCH_PERSON = `${MODULE_NAME}/FETCH_PERSON`

// Pending handling is a bit sensitive in the profile banner. It needs to be
// undefined (not null) unless/until we arrive at a better solution. See
// definition of `pending` in connector.
// TODO: the sort of thing that could be made into a generic action creator
export function clearFetchPersonPending () {
  return {
    type: CLEAR_FETCH_PERSON_PENDING
  }
}

export function fetchPerson (id) {
  return {
    type: FETCH_PERSON,
    graphql: {
      query: `query PersonDetails ($id: ID) {
        person (id: $id) {
          id
          name
          avatarUrl
          bannerUrl
          bio
          twitterName
          linkedinUrl
          facebookUrl
          url
          tagline
          location
          messageThreadId
          memberships {
            id
            role
            hasModeratorRole
            community {
              id
              name
              slug
              memberCount
            }
          }
          skills (first: 100) {
            total
            hasMore
            items {
              id
              name
            }
          }
        }
      }`,
      variables: { id }
    },
    meta: {
      afterInteractions: true,
      extractModel: 'Person'
    }
  }
}

export const getPerson = ormCreateSelector(
  orm,
  state => state.orm,
  (_, { id }) => id,
  (session, id) => {
    const person = session.Person.safeGet({id})
    if (!person) return null
    return {
      ...person.ref,
      skills: person.skills.toModelArray(),
      memberships: person.memberships.toModelArray()
    }
  }
)
