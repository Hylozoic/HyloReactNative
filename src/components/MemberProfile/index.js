import { withNavigationFocus } from 'react-navigation-is-focused-hoc'

import component from './MemberProfile'
import connector from './MemberProfile.connector'
export default withNavigationFocus(connector(component))
