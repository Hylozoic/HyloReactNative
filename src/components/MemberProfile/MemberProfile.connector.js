import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { get } from 'lodash/fp'
import { getPerson, fetchPerson } from './MemberProfile.store'
import blockUser from '../../store/actions/blockUser'
import makeGoToCommunity from '../../store/actions/makeGoToCommunity'
import getMe from '../../store/selectors/getMe'
import getBlockedUsers from '../../store/selectors/getBlockedUsers'
import updateUserSettings from '../../store/actions/updateUserSettings'
import { mapWhenFocused, mergeWhenFocused } from 'util/connector'

export function mapStateToProps (state, props) {
  const id = get('navigation.state.params.id', props)
  const editing = get('navigation.state.params.editing', props)

  const isBlocked = !!getBlockedUsers(state).find(i => get('id', i) === id)
  const person = getPerson(state, {id})
  const goToDetails = () => props.navigation.navigate('MemberDetails', { id })
  const goToEdit = () => props.navigation.navigate('MemberDetails', { id, editing: true })
  const goToSkills = () => props.navigation.navigate('MemberSkillEditor', { id } )
  const currentUser = getMe(state, props)
  const skills = person ? person.skills : []
  const isMe = Number(get('id', currentUser)) === Number(id)
  const navigation = props.navigation

  return {
    isBlocked,
    id,
    editing,
    person,
    currentUser,
    skills,
    goToDetails,
    goToEdit,
    goToSkills,
    isMe,
    navigation
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    goToCommunity: makeGoToCommunity(dispatch, props.navigation),
    ...bindActionCreators({
      fetchPerson,
      updateUserSettings,
      blockUser
    }, dispatch)
  }
}

export function makeOnPressMessages (currentUser, person, navigation) {
  if (!person || currentUser.id === person.id) return () => navigation.navigate('ThreadList')
  const { messageThreadId } = person
  if (messageThreadId) return () => navigation.navigate('Thread', { id: messageThreadId })
  return () => navigation.navigate('NewMessage',{ participants: [person.id] })
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { id, currentUser, person, isMe } = stateProps
  const { navigation } = ownProps

  const fetchPerson = () => dispatchProps.fetchPerson(id)

  const canFlag = currentUser && id && currentUser.id !== id

  const onPressMessages = makeOnPressMessages(currentUser, person, navigation)

  // only call updateUserSettings if isMe (otherwise when you visit another
  // user and then press the back button, you will update your profile with their
  // settings)
  const updateUserSettings = isMe ? dispatchProps.updateUserSettings : () => {}

  const goToMemberProfile = () => navigation.navigate('MemberProfile', { id })

  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    updateUserSettings,
    canFlag,
    fetchPerson,
    onPressMessages,
    goToMemberProfile
  }
}

export default connect(
  mapWhenFocused(mapStateToProps),
  mapWhenFocused(mapDispatchToProps),
  mergeWhenFocused(mergeProps)
)
