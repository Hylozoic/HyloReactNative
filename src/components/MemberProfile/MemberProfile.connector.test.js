import { mapStateToProps, mapDispatchToProps, makeOnPressMessages, mergeProps } from './MemberProfile.connector'

describe('mapStateToProps', () => {
  it('maps the state to the props', () => {
    const id = 123
    const navigation = {
      state: {params: {id}},
      navigate: jest.fn()
    }
    const props = mapStateToProps({}, {navigation})
    expect(props).toMatchSnapshot()
    props.goToDetails()
    expect(navigation.navigate).toHaveBeenCalledWith('MemberDetails', {id})
    navigation.navigate.mockClear()
    props.goToEdit()
    expect(navigation.navigate).toHaveBeenCalledWith('MemberDetails', {id, editing: true})
  })
})

describe('mapDispatchToProps', () => {
  it('returns the right keys', () => {
    const props = {
      navigation: {navigate: jest.fn()}
    }
    expect(mapDispatchToProps({}, props)).toMatchSnapshot()
  })
})

describe('makeOnPressMessages', () => {
  var navigation

  beforeEach(() => {
    navigation = {
      navigate: jest.fn()
    }
  })

  it("navigates to thread list when it's your profile", () => {
    const onPressMessages = makeOnPressMessages({id: 1}, {id: 1}, navigation)
    onPressMessages()
    expect(navigation.navigate).toHaveBeenCalledWith('ThreadList')
  })

  it('navigates to existing thread when there is one', () => {
    const messageThreadId = 99
    const onPressMessages = makeOnPressMessages({id: 1}, {id: 2, messageThreadId}, navigation)
    onPressMessages()
    expect(navigation.navigate).toHaveBeenCalledWith('Thread', {id: messageThreadId})
  })

  it('navigates to new message page when there is no existing thread', () => {
    const otherPersonId = 2
    const onPressMessages = makeOnPressMessages({id: 1}, {id: otherPersonId}, navigation)
    onPressMessages()
    expect(navigation.navigate).toHaveBeenCalledWith('NewMessage', {participants: [otherPersonId]})
  })
})

describe('mergeProps', () => {
  it('returns the right keys', () => {
    const id = 78
    const stateProps = {
      currentUser: {id: 1},
      person: {id},
      id
    }
    const dispatchProps = {
      fetchPerson: jest.fn()
    }
    const ownProps = {
      navigation: {
        navigate: () => {}
      }
    }
    const mergedProps = mergeProps(stateProps, dispatchProps, ownProps)
    expect(mergedProps).toMatchSnapshot()
    mergedProps.fetchPerson()
    expect(dispatchProps.fetchPerson).toHaveBeenCalledWith(id)
  })
})
