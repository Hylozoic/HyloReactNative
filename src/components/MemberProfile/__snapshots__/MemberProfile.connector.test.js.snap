// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps returns the right keys 1`] = `
Object {
  "fetchPerson": [Function],
  "goToCommunity": [Function],
  "updateUserSettings": [Function],
}
`;

exports[`mapStateToProps maps the state to the props 1`] = `
Object {
  "currentUser": undefined,
  "editing": undefined,
  "goToDetails": [Function],
  "goToEdit": [Function],
  "goToSkills": [Function],
  "id": 123,
  "isMe": false,
  "person": null,
}
`;

exports[`mergeProps returns the right keys 1`] = `
Object {
  "canFlag": true,
  "currentUser": Object {
    "id": 1,
  },
  "fetchPerson": [Function],
  "id": 78,
  "navigation": Object {
    "navigate": [Function],
  },
  "onPressMessages": [Function],
  "person": Object {
    "id": 78,
  },
}
`;
